Add caption property to ListItem

DanielMcInnes opened the pull request:
Fixes #1822

blammit commented on components/listitems/ListLink.qml:
```diff
@@ -56,15 +52,13 @@ ListItem {
-			visible: root.mode === VenusOS.ListLink_Mode_QRCode
-		}
-	]
+	caption.horizontalAlignment: Text.AlignLeft
```
The Text.AlignLeft is redundant?

blammit commented on components/listitems/core/ListItem.qml:
```diff
@@ -104,6 +105,13 @@ Item {
+			PrimaryListLabel {
+				id: caption
+
+				width: Math.min(caption.implicitWidth, root.width - root.leftPadding - root.rightPadding)
```
This can be `width: parent.width`, otherwise the text spills onto the right-hand content:

![image](https://github.com/user-attachments/assets/cab14e3a-13c3-4a65-ba60-75f67effda90)

blammit reviewed (commented):
Can this be reduced to just having a single `caption` property in ListItem? 

- It looks like Theme.color_listItem_secondaryText is always the preferred color, so it doesn't need customization
- I tried setting topPadding=0 and bottomPadding=0 by default in the caption label, and this still looked fine for ListLink, which seems to be the only case where these values aren't initialized to something else
- Components can just set caption to an empty string when it shouldn't be displayed, e.g. in the PageGensetModel.qml case

Also, maybe ListItem can just use a generic Label and style it accordingly, instead of using PrimaryListLabel and giving it a secondary font color. Since it is placing the label inside the `bottomContent`, it can just set `visible` instead of `preferredVisible` (which is intended to be used by the root item, rather than children).

DanielMcInnes commented:
> Can this be reduced to just having a single `caption` text property in ListItem?

If we added a `caption` text property to ListItem, we would also have to add `captionVisible`, `captionColor`, and arguably `captionTopPadding` and `captionBottomPadding`. Would you prefer it like that?

> 
>     * It looks like Theme.color_listItem_secondaryText is always the preferred color, so it doesn't need customization

I just changed the default color to `Theme.color_font_secondary`. It is overridden in PageGensetModel, ListLink, PageSettingsFroniusInverter and PageSettingsDemo

> 
>     * I tried setting topPadding=0 and bottomPadding=0 by default in the caption label, and this still looked fine for ListLink, which seems to be the only case where these values aren't initialized to something else

PageGensetModel, PageSettingsCanbus, and PageSettingsFroniusInverter also use `caption` without initializing topPadding and bottomPadding

> 
>     * Components can just set caption to an empty string when it shouldn't be displayed, e.g. in the PageGensetModel.qml case


Updated as suggested


> 
> Also, maybe ListItem can just use a generic Label and style it accordingly, instead of using PrimaryListLabel and giving it a secondary font color. 

Updated as suggested

>
> Since it is placing the label inside the `bottomContent`, it can just set `visible` instead of `preferredVisible` (which is intended to be used by the root item, rather than children).

Updated as suggested

blammit commented:
> PageGensetModel, PageSettingsCanbus, and PageSettingsFroniusInverter also use caption without initializing topPadding and bottomPadding

They seem to look okay with topPadding=0 and bottomPadding=0. For example, here is the PageGensetModel:

![image](https://github.com/user-attachments/assets/98766e28-d35f-4a1d-a3af-e7edc5139a7f)

I'm thinking, how about we just use topPadding=0 and bottomPadding=0 everywhere for visual consistency?

>  just changed the default color to Theme.color_font_secondary. It is overridden in PageGensetModel, ListLink, PageSettingsFroniusInverter and PageSettingsDemo

Let's drop the color customizations, and just use the default secondary colour everywhere for captions, so that it is visually consistent.

> If we added a caption text property to ListItem, we would also have to add captionVisible, captionColor, and arguably captionTopPadding and captionBottomPadding. Would you prefer it like that?

No, that would be annoying :-(

Ideally the captions look the same everywhere, and ideally just a `caption` string property would be sufficient. If we really need customization then perhaps we could add a `captionLabel` property that is aliased to the ListItem label. Are there any cases where we need customizations, if we align the padding and colouring in the current use cases?

DanielMcInnes commented on components/PageGensetModel.qml:
```diff
@@ -361,11 +361,9 @@ ObjectModel {
-								text: qsTrId("genset_bms_control_enabled_automatically")
-								preferredVisible: bmsControlled.dataItem.value === 1
-							}
+							caption.color: Theme.color_font_primary
```
dropped

DanielMcInnes commented:
> > PageGensetModel, PageSettingsCanbus, and PageSettingsFroniusInverter also use caption without initializing topPadding and bottomPadding
> 
> They seem to look okay with topPadding=0 and bottomPadding=0. For example, here is the PageGensetModel:
> 
> ![image](https://private-user-images.githubusercontent.com/1260099/411845436-98766e28-d35f-4a1d-a3af-e7edc5139a7f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyNTY0NDgsIm5iZiI6MTczOTI1NjE0OCwicGF0aCI6Ii8xMjYwMDk5LzQxMTg0NTQzNi05ODc2NmUyOC1kMzVmLTRhMWQtYTNhZi1lN2VkYzUxMzlhN2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTFUMDY0MjI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjY0MjA1YzljYzM5ZmIwMWRmYThkNjZhNWE4MmUxMmMxNWZmZTUwMmNmNGIyNzA2ZGVkYjcyOTcyMGNjZjVhMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.6Wu7JhWi4RQ8T-d-pZ0hggl6oNsKLIL4OHvrAupzXqE)
> 
> I'm thinking, how about we just use topPadding=0 and bottomPadding=0 everywhere for visual consistency?
> 
> > just changed the default color to Theme.color_font_secondary. It is overridden in PageGensetModel, ListLink, PageSettingsFroniusInverter and PageSettingsDemo
> 
> Let's drop the color customizations, and just use the default secondary colour everywhere for captions, so that it is visually consistent.
> 
> > If we added a caption text property to ListItem, we would also have to add captionVisible, captionColor, and arguably captionTopPadding and captionBottomPadding. Would you prefer it like that?
> 
> No, that would be annoying :-(
> 
> Ideally the captions look the same everywhere, and ideally just a `caption` string property would be sufficient. If we really need customization then perhaps we could add a `captionLabel` property that is aliased to the ListItem label. Are there any cases where we need customizations, if we align the padding and colouring in the current use cases?

No, there are no cases where we need customizations, if we align the padding and colouring in the current use cases. Updated as suggested.

blammit commented on components/listitems/core/ListItem.qml:
```diff
@@ -112,6 +113,20 @@ Item {
 			Layout.bottomMargin: Theme.geometry_listItem_content_verticalMargin
+
+			Label {
+				id: caption
```
id can be removed

DanielMcInnes replied:
removed

blammit commented on components/listitems/core/ListItem.qml:
```diff
@@ -14,6 +14,7 @@ Item {
 	property alias content: content
 	property alias bottomContent: bottomContent
 	property alias bottomContentChildren: bottomContent.children
+	property string caption: ""
```
"" initialization is not needed?

DanielMcInnes replied:
thanks, updated

blammit reviewed (approved):
Looks great with the clean-up :)
